plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,849 @@
+/*
+ * Copyright (C) 2023 Crag Wang <crag.wang@dell.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fcntl.h>
+#include <glib/gstdio.h>
+#include <linux/i2c-dev.h>
+
+#include "fu-mediatek-scaler-device.h"
+
+#define DDC_DISPLAY_I2C_ADDR	 0x6E
+#define DDC_HOST_I2C_ADDR	 0x51
+#define DDC_HOST_I2C_ADDR_CHKSUM 0x50
+#define DDC_DATA_LEN_DFT	 0x80
+#define DDC_OPCODE_MTK_REQ	 0xCC
+#define DDC_VCP_CODE_PRIORITY    0x90
+#define DDC_VCP_CODE_UPDATE_PREP 0xF2
+#define DDC_VCP_CODE_UPDATE_ACK	 0xF3
+#define DDC_VCP_CODE_SET_DATA	 0xF4
+#define DDC_VCP_CODE_GET_STAGED  0xF5
+#define DDC_VCP_CODE_SET_DATA_FF	0xF6
+#define DDC_VCP_CODE_COMMIT_FW   0xF7
+#define DDC_VCP_CODE_GET_ISP_MODE       0xF8
+#define DDC_VCP_CODE_REBOOT	 0xFB
+#define DDC_VCP_CODE_SUM	 0xFE
+#define DDC_VCP_CODE_VERSION	 0xFF
+#define FLASH_FIRMWARE_SIZE	 0x1E848
+#define DDC_DATA_FRAGEMENT_SIZE	 0x0B   /* 11 bytes for each DDC write */
+#define DDC_DATA_PAGE_SIZE		0x1000 /* 4K bytes for each block page */
+#define DDC_RW_MAX_RETRY_CNT     10
+
+/* timeout duration in ms to i2c-dev operation */
+#define FU_MEDIATEK_SCALER_DEVICE_IOCTL_TIMEOUT 5000
+
+/* delay time in ms before sending the next message */
+#define FU_MEDIATEK_SCALER_DDC_MSG_DELAY 50
+
+/* interval in ms between the poll to check device status */
+#define FU_MEDIATEK_SCALER_DEVICE_POLL_INTERVAL 1000
+
+typedef enum {
+	DDCCI_PRI_NORMAL,
+	DDCCI_PRI_UP,
+} FuMediatekScalerSetPriority;
+
+struct _FuMediatekScalerDevice {
+	FuUdevDevice parent_instance;
+	FuUdevDevice *i2c_dev;
+	gchar *i2c_dev_name;
+	gchar *dp_aux_dev_name;
+};
+
+G_DEFINE_TYPE(FuMediatekScalerDevice, fu_mediatek_scaler_device, FU_TYPE_UDEV_DEVICE)
+
+static gboolean
+fu_mediatek_scaler_ensure_device_address(FuMediatekScalerDevice *self,
+					 guint8 address,
+					 GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+
+	if (!fu_udev_device_ioctl(self->i2c_dev,
+				  I2C_SLAVE,
+				  (guint8 *)(guintptr)address,
+				  NULL,
+				  FU_MEDIATEK_SCALER_DEVICE_IOCTL_TIMEOUT,
+				  &error_local)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "Failed to set address on '%s' for '%s': %s",
+			    self->i2c_dev_name,
+			    self->dp_aux_dev_name,
+			    (error_local != NULL ? error_local->message : "unknown"));
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_set_i2c_dev(FuMediatekScalerDevice *self,
+				      const GPtrArray *i2c_devices,
```

Comment:
I suggest renaming this to something different, maybe `devices`?  Otherwise it's confusing to have `i2c_devices` and `i2c_devs` both so similar in the same function.

Reply:
Fixed